### PR TITLE
Typed api.storeResult

### PR DESCRIPTION
**Pull request by mischnic**
Not sure if this is actually a good idea, but I find it very confusing and error-prone that these two cases typecheck just fine
```js
	// Does returns undefined if uncached, but returns a number if cached
    let requestB: void = await tracker.runRequest({
      id: 'abc',
      type: 'mock_request',
      run: async ({api}) => {
        api.storeResult(123);
      },
      input: null,
    });
```
```js
	// Actually returns undefined if cached
    let requestB: string = await tracker.runRequest({
      id: 'abc',
      type: 'mock_request',
      run: async ({api}) => {
        return 'b'
      },
      input: null,
    });
```

### CODE REVIEW

File: packages/core/core/src/requests/DevDepRequest.js [hunk 1]
```diff
@@ -163,11 +165,22 @@ export function invalidateDevDeps(
...
+
+export async function runDevDepRequest<TResult>(
+  api: RunAPI<TResult>,
   devDepRequest: DevDepRequest,
 ) {
-  await api.runRequest<null, void>({
+  await api.runRequest<null, DevDepRequestResult | void>({
     id: 'dev_dep_request:' + devDepRequest.specifier + ':' + devDepRequest.hash,
     type: 'dev_dep_request',
     run: ({api}) => {
```
Comment (mischnic) [inline packages/core/core/src/requests/DevDepRequest.js:186]: Also strange: this function has a `api.storeResult` call but AFAICT that value is actually never used/read anywhere

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>Kitchen Sink ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 1.74s | -57.00ms |
| Cached | 384.00ms | -1.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.7cdb0fad.webp | 102.94kb | +0.00b | 106.00ms | -172.00ms 🚀 |
| dist/legacy/parcel.7cdb0fad.webp | 102.94kb | +0.00b | 107.00ms | -172.00ms 🚀 |
| dist/modern/parcel.7cdb0fad.webp | 102.94kb | +0.00b | 106.00ms | -173.00ms 🚀 |
| dist/legacy/index.b8ae99ba.css | 94.00b | +0.00b | 301.00ms | +18.00ms ⚠️ |
| dist/modern/index.31cedca9.css | 94.00b | +0.00b | 300.00ms | +17.00ms ⚠️ |

</p></details>

<details><summary>React HackerNews ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 11.26s | -48.00ms |
| Cached | 508.00ms | -8.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

*No bundle changes detected.*

</p></details>

<details><summary>AtlasKit Editor ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 1.94m | -1.24s |
| Cached | 2.57s | +57.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

*No bundle changes detected.*

</p></details>

<details><summary>Three.js ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 8.15s | -77.00ms |
| Cached | 307.00ms | -8.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

*No bundle changes detected.*

</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/353133465122636376)**

File: packages/core/core/src/requests/DevDepRequest.js (hunk 1 above)
Comment (devongovett) [reply packages/core/core/src/requests/DevDepRequest.js:186]: does removing it break something?

Comment (mischnic) [reply packages/core/core/src/requests/DevDepRequest.js:186]: Nevermind, the result is loaded here
https://github.com/parcel-bundler/parcel/blob/ea61ad9dc295133eeb50cce3e1665f4dfc08b632/packages/core/core/src/requests/AssetRequest.js#L80-L90

**Review by devongovett — APPROVED** [review]

**Comment by mischnic** [comment]
(One unsoundness does remain: `runRequest()` can still return undefined if the request didn't do `.storeResult()`, but enforcing that isn't possible with Flow, and the alternative of making the result always `?TResult` is very unergonomic.)

**Pull request merged**